### PR TITLE
Replace request package #20

### PR DESCRIPTION
**Pull request by lvenier**
simple draft to replace request by fetch.

Deploy function is returning http 500. I would need some logs from backend to understand the reason.
Clean up to be done when issue is resolved as code can be improved/refactored

### CODE REVIEW

**Comment by martinlevesque** [comment]
Thanks for the PL, will look at it over the week.

Will need to test all commands also.

**Comment by martinlevesque** [comment]
Just tested your branch.

The issue is when trying to transfer the synched files, when requesting 

POST /instances/sitename/sendCompressedFile

Normally our rails application interprets the sent file properly, so "param file" ` #<ActionDispatch::Http::UploadedFile:0x00007ff100842f08>`

However in your branch, param file sent is not interpreted, and following object is received

`{"_readableState"=>{"objectMode"=>false, "highWaterMark"=>65536, "buffer"=>{"head"=>nil, "tail"=>nil, "length"=>0}, "length"=>0, "pipes"=>nil, "pipesCount"=>0, "flowing"=>nil, "ended"=>false, "endEmitted"=>false, "reading"=>false, "sync"=>true, "needReadable"=>false, "emittedReadable"=>false, "readableListening"=>false, "resumeScheduled"=>false, "paused"=>true, "emitClose"=>false, "autoDestroy"=>false, "destroyed"=>false, "defaultEncoding"=>"utf8", "awaitDrain"=>0, "readingMore"=>false, "decoder"=>nil, "encoding"=>nil}, "readable"=>true, "_events"=>{}, "_eventsCount"=>1, "path"=>"cc9f0ba25bdcd7d677cd60739ce4c910.zip", "fd"=>nil, "flags"=>"r", "mode"=>438, "end"=>nil, "autoClose"=>true, "bytesRead"=>0, "closed"=>false}`

The same POST content should be sent here (original version), if you could compare and adapt your version it would be great

https://github.com/openode-io/openode-cli/blob/4484935ad2970e4ec1df46f43647d8c1ebaacdbf/modules/deploy.js#L91

**Comment by lvenier** [comment]
@martinlevesque : ok got it. You are using form-data. I have updated the PR and looks to be working.
We need a new require : form-data. It was used by request but node-fetch does not have it within its code.

**Comment by martinlevesque** [comment]
with the new version, openode sync seems failing:

```
ActionController::BadRequest (Invalid request parameters: invalid %-encoding ("file"; filename="cc9f0ba25bdcd7d677cd60739ce4c910.zip"
Content-Type: application/zip
```

**Comment by lvenier** [comment]
I have just ran the command and did not get any issues : 
```
# node /tmp/openode-cli/index.js sync
 ⠹ Processing for usa
⠏ Sending files (trial 1) - file=./Dockerfile size=167 operation=C
⠇ 1/1
[
  {
    result: 'success',
    details: '1 change(s), 0 deletion(s)',
    location: 'usa'
  }
]
```
can you pls show to reproduce that issue ?

**Comment by martinlevesque** [comment]
Doing the following commands

```
node ../lvenier-openode-cli/index.js erase-all
node ../lvenier-openode-cli/index.js sync
```

leads to

```
[Sun May 24 2020 09:42:28 GMT-0400 (Eastern Daylight Time)] -  [+] Authentication valid.
⠋ Processing for eu
Sending files (trial 1) - file=./Dockerfile size=607 operation=C
Sending files (trial 1) - file=./Dockerfile.bak size=396 operation=C
Sending files (trial 1) - file=./README.md size=31 operation=C
Sending files (trial 1) - file=./certs/tls.crt size=1814 operation=C
Sending files (trial 1) - file=./certs/tls.key size=3272 operation=C
Sending files (trial 1) - file=./docker-compose-bak.yml size=129 operation=C
Sending files (trial 1) - file=./index.html size=35093 operation=C
Sending files (trial 1) - file=./package-lock.json size=74 operation=C
Sending files (trial 1) - file=./package.json size=285 operation=C
Sending files (trial 1) - file=./server.js size=470 operation=C
Sending files (trial 1) - file=./what size=12 operation=C
⠴ 
... failed, retrying
Sending files (trial 2) - file=./Dockerfile size=607 operation=C
Sending files (trial 2) - file=./Dockerfile.bak size=396 operation=C
Sending files (trial 2) - file=./README.md size=31 operation=C
Sending files (trial 2) - file=./certs/tls.crt size=1814 operation=C
Sending files (trial 2) - file=./certs/tls.key size=3272 operation=C
Sending files (trial 2) - file=./docker-compose-bak.yml size=129 operation=C
Sending files (trial 2) - file=./index.html size=35093 operation=C
Sending files (trial 2) - file=./package-lock.json size=74 operation=C
Sending files (trial 2) - file=./package.json size=285 operation=C
Sending files (trial 2) - file=./server.js size=470 operation=C
Sending files (trial 2) - file=./what size=12 operation=C
⠦ 
... failed, retrying
Sending files (trial 3) - file=./Dockerfile size=607 operation=C
Sending files (trial 3) - file=./Dockerfile.bak size=396 operation=C
Sending files (trial 3) - file=./README.md size=31 operation=C
Sending files (trial 3) - file=./certs/tls.crt size=1814 operation=C
Sending files (trial 3) - file=./certs/tls.key size=3272 operation=C
Sending files (trial 3) - file=./docker-compose-bak.yml size=129 operation=C
Sending files (trial 3) - file=./index.html size=35093 operation=C
Sending files (trial 3) - file=./package-lock.json size=74 operation=C
Sending files (trial 3) - file=./package.json size=285 operation=C
Sending files (trial 3) - file=./server.js size=470 operation=C
Sending files (trial 3) - file=./what size=12 operation=C
⠧ 
Error: Failed sending
    at execSyncFiles (/home/martin/works/lvenier-openode-cli/modules/deploy.js:253:19)
[ { Error: Failed sending
      at execSyncFiles (/home/martin/works/lvenier-openode-cli/modules/deploy.js:253:19) location: 'eu' } ]
```

on the server side:

```
Started POST "/instances/newnewnewtest345/sendCompressedFile" for 127.0.0.1 at 2020-05-24 09:42:29 -0400
  
ActionController::BadRequest (Invalid request parameters: invalid %-encoding ("file"; filename="cc9f0ba25bdcd7d677cd60739ce4c910.zip"
Content-Type: application/zip

#�cP
    ./.openode2�A
� л̺M��t�P�BDS�PDt�j��M����ls[�e�
                                s���'����b����SQ�k?��j��A��׊��N�
                                                                PT��}U_��P
                                                                          ./Dockerfile}�IO�0���O��J���	U�CB]���c��[7�l���;�����d���73O�l�ٿ�˘����|����K����Ǭ%d<��|��=�����;8��)|֦`Z�*b ��I���SH^�t�
���)���                             �d����K�J��PJx�
�BD���
      �W��R�J����L���M�/�w1X���r��h�G��JPJ�����N�9��X�H���<�N�,�k���
                                                                    g�/����G���8MÝ�Yf�5?�7�+�%�-�M��<p(h�^��G���
                                                                                                                V��d$5醼Y
                                                                                                                         ��C)):
  
/home/martin/.rvm/rubies/ruby-2.7.1/lib/ruby/2.7.0/uri/common.rb:387:in `decode_www_form_component'
rack (2.2.2) lib/rack/utils.rb:54:in `unescape'

```

I can debug it over the week if not already fixed

**Comment by lvenier** [comment]
weird : 
```
#  node /tmp/openode-cli/index.js erase-all
[Sun May 24 2020 10:04:11 GMT-0400 (Eastern Daylight Time)] -  [+] Authentication valid.
⠹ Processing for usa
[ { result: 'success', location: 'usa' } ]
# node /tmp/openode-cli/index.js sync
⠸ 244/244
[
  {
    result: 'success',
    details: '244 change(s), 0 deletion(s)',
    location: 'usa'
  }
]
```

**Comment by martinlevesque** [comment]
Ok I see it's working with the live API.
It's not working however with my dev API instance. Will look what could be the issue.

**Comment by lvenier** [comment]
@martinlevesque: I do not encodeURI the url. Might be the reason....

**Comment by martinlevesque** [comment]
Just made it work by using NGINX reverse proxyfied as we do in prod - so there is a certain difference in what the client is sending, which is somehow fixed by NGINX - more research needed :-O.

**Comment by lvenier** [comment]
Good to know. Surprising though. 😋 If you output a http request from a direct call and ine through nginx you should see the difference. I'd love to know...

**Comment by martinlevesque** [comment]
If it's something possible from your side: could you prepare a second upload function (the other one using node-fetch commented) and use **axios** for uploading only (for now).

axios looks a solid alternative ( https://blog.logrocket.com/axios-or-fetch-api/ ). If there is no issue with uploading using **axios** it would be a better choice, if both have the same issue with our server then we can keep **node-fetch**.

**Comment by lvenier** [comment]
@martinlevesque Yes I can but I would prefer to understand whats going on.
Would you be able to print out the http header when it fails with the new version and the header when it works with master ?
with both we can see the difference....

**Comment by martinlevesque** [comment]
Sure, find below, using wireshark listening on the server port

lvenier branch:

```
POST /instances/newnewnewtest345/sendCompressedFile HTTP/1.1
x-auth-token: XXXX
User-Agent: express
Content-Type: multipart/form-data;boundary=--------------------------725789911057008519540553
Accept: */*
Accept-Encoding: gzip,deflate
Connection: close
Host: localhost:3000
Transfer-Encoding: chunked
```


openode-cli master:

```
POST /instances/newnewnewtest345/sendCompressedFile HTTP/1.1
x-auth-token: XXXX
host: localhost:3000
content-type: multipart/form-data; boundary=--------------------------016126237501673073830437
content-length: 17024
Connection: close
```

Ok I believe it can work by specifying the content length explicitly.
In req.js upload(path, params, config, url = null, length = null)
add a length, then the headers can be

```
const headers = {
      "x-auth-token": config.token,
      'User-Agent': 'express'
    }

    if (length) {
      headers['Content-Length'] = length
    }
```

The formdata length can be retrieved by using getLength, https://github.com/form-data/form-data#integer-getlength-function-callback-

Tested it by specifying the hard coded length value and worked. If you could make this change it would be great.

**Comment by lvenier** [comment]
@martinlevesque : Good catch ! you defintely have found the reason. It makes completely sense.
It is hard for me to test as it is working. I have made a blind update of the PR with a promise on Getlength. Not sure if it's fixed it :(
Let me know.

PS : Can you provide a test url I can make some tests ?

**Comment by martinlevesque** [comment]
Seems working well.

> PS : Can you provide a test url I can make some tests ?

No we do not really have a "staging" environment.

I will review the code, then add some automated tests (would like to add more test coverage soon), then test most commands before publishing it.

File: modules/deploy.js [hunk 1]
```diff
@@ -63,127 +60,60 @@ async function localFilesListing(dir, files2Ignore, firstLevel = false) {
...
+async function sendCompressedFile(file, config, options) {
+  let formData = new FormData()
+  let file2Upload = fs.createReadStream(file);
+  formData.append('file', file2Upload)
+  formData.append('info', JSON.stringify({
+      "path": file
+    }))
+  formData.append('version', config.version)
+  formData.append('location_str_id', options.location_str_id)
+  const Length = await new Promise((resolve, reject) => {
```
Comment (martinlevesque) [inline modules/deploy.js:79]: could be good to make a separate function

File: modules/deploy.js [hunk 2]
```diff
@@ -63,127 +60,60 @@ async function localFilesListing(dir, files2Ignore, firstLevel = false) {
...
+    formData.getLength(function(err, length) {
+      resolve(length)
```
Comment (martinlevesque) [inline modules/deploy.js:81]: handle the error:

`if (err) { reject(err) } else { ... }`

File: modules/deploy.js (hunk 1 above)
Comment (martinlevesque) [inline modules/deploy.js:79]: Use camelcase, Length -> length

File: modules/deploy.js [hunk 3]
```diff
@@ -63,127 +60,60 @@ async function localFilesListing(dir, files2Ignore, firstLevel = false) {
...
+    })
+  })
+  return fetch.upload('instances/' + config.site_name + "/sendCompressedFile", formData, config, null, Length)
```
Comment (martinlevesque) [inline modules/deploy.js:84]: better to stringify with:

`instances/${config.site_name}/sendCompressedFile`

File: modules/deploy.js [hunk 4]
```diff
@@ -63,127 +60,60 @@ async function localFilesListing(dir, files2Ignore, firstLevel = false) {
...
-      formData: formData
-    }, function optionalCallback(err, httpResponse, body) {
-      if (err || httpResponse.statusCode != 200) {
-        reject(err);
-      } else {
-        resolve(body);
-      }
-    });
-  });
+async function sendFile(file, config, options) {
```
Comment (martinlevesque) [inline modules/deploy.js:87]: It's no more used, remove it.

and remove the test test/deploy_test.js:96

File: modules/instance.js [hunk 5]
```diff
@@ -1,102 +1,37 @@
-const cliConfs = require("./cliConfs");
-const request = require("request");
+const fetch = require("./req");
```
Comment (martinlevesque) [inline modules/instance.js:1]: fetch is confusing with node-fetch

File: modules/instance.js [hunk 6]
```diff
@@ -1,102 +1,37 @@
...
-    }, function optionalCallback(err, httpResponse, body) {
-      if (err || httpResponse.statusCode != 200) {
-        log.err(body); // show error
-        resolve(null);
-      } else {
-        resolve(body); // returns the created website info
-      }
-    });
-  });
+    if ( ! opts.sitename || opts.sitename == "") resolve(false);
```
Comment (martinlevesque) [inline modules/instance.js:14]: probably better to do

```
if ( ! opts.sitename || opts.sitename == "") {
  return resolve(false);
}
```

File: modules/instanceRequest.js [hunk 7]
```diff
@@ -1,92 +1,20 @@
-
-const request = require("request");
-const cliConfs = require("./cliConfs");
+const fetch = require("./req");
```
Comment (martinlevesque) [inline modules/instanceRequest.js:1]: rename

File: modules/instanceRequest.js [hunk 8]
```diff
@@ -1,92 +1,20 @@
...
 function getOp(operation, sitename, config, options = {}) {
-  return new Promise((resolve, reject) => {
-
-    if (!sitename || sitename == "") {
-      reject({});
-    }
-
-    let url = cliConfs.getApiUrl() + 'instances/' + sitename + "/" + operation + "?";
-
+    if (!sitename || sitename == "") return {}
```
Comment (martinlevesque) [inline modules/instanceRequest.js:5]: it's not returning a promise

File: modules/instanceRequest.js [hunk 9]
```diff
@@ -1,92 +1,20 @@
...
-      form: form
-    }, function optionalCallback(err, httpResponse, body) {
-      if (err || httpResponse.statusCode != 200) {
-        reject(body);
-      } else {
-        resolve(body);
-      }
-    });
-  });
+    if (!sitename || sitename == "") return {}
```
Comment (martinlevesque) [inline modules/instanceRequest.js:11]: it's not returning a promise

File: modules/req.js [hunk 10]
```diff
@@ -1,49 +1,129 @@
 const cliConfs = require("./cliConfs");
-const request = require("request");
```
Comment (martinlevesque) [inline modules/req.js:2]: great

File: modules/req.js [hunk 11]
```diff
@@ -1,49 +1,129 @@
...
+        method: 'POST',
+        headers: {
+          "content-type": "application/json",
+          "x-auth-token": config.token,
+          'User-Agent': 'express'
+        },
+        body: JSON.stringify(params)
+      })
+      .then(response => {
+        if (response.ok) {
```
Comment (martinlevesque) [inline modules/req.js:48]: this part is repeated multiple times, refactor by making a function and calling it

File: modules/req.js [hunk 12]
```diff
@@ -1,49 +1,129 @@
...
+        },
+        body: params
+      })
+      .then(response => {
+        if (response.ok) {
+          response.text().then(function (data) {
+            resolve(data);
+          })
+        } else {
+          response.text().then(function (data) {
```
Comment (martinlevesque) [inline modules/req.js:84]: it's .text() only when uploading?

File: modules/templates.js [hunk 13]
```diff
@@ -1,55 +1,14 @@
-const cliConfs = require("./cliConfs");
-const request = require("request");
+const fetch = require("./req");
```
Comment (martinlevesque) [inline modules/templates.js:1]: rename

File: package.json [hunk 14]
```diff
@@ -39,11 +39,12 @@
     "asciify": "^1.3.5",
     "commander": "^5.1.0",
     "compare-version": "^0.1.2",
+    "form-data": "*",
```
Comment (martinlevesque) [inline package.json:42]: please put the expected version

File: package.json [hunk 15]
```diff
@@ -39,11 +39,12 @@
...
     "is-glob": "^4.0.1",
     "lodash": "^4.17.15",
+    "node-fetch": "*",
```
Comment (martinlevesque) [inline package.json:45]: please put the expected version

File: modules/req.js (hunk 12 above)
Comment (lvenier) [reply modules/req.js:84]: I would expect text body response in upload form-data content.
can be changed to json if you confirm the backend is a json doc.

**Comment by lvenier** [comment]
@martinlevesque: Ive made  couple of requested changes.
Best would be to have sonarCube or similar to check duplicated/useless code.

File: modules/deploy.js (hunk 3 above)
Comment (lvenier) [reply modules/deploy.js]: best would be to change it every where.... :(

**Comment by martinlevesque** [comment]
I'll merge it to master. will add some unit tests and do more manual tests.
Any feedback welcome.

**Pull request merged**